src/audit-log/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+//! new version here can be made. This avoids needing to poke at the data
+//! when reading it to determine what it means and should have full backward
+//! compatibility. This is its own crate so that production and consumption can
+//! be in different processes and production is not allowed to specify private
+//! data structures unknown to the reader.
+
+use std::{collections::BTreeSet, num::NonZeroUsize};
+
+use bytesize::ByteSize;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+/// New version variants should be added if fields need to be added, changed, or removed.
+#[derive(Serialize, Deserialize)]
+pub enum VersionedEvent {
+    V1(EventV1),
+}
+
+impl VersionedEvent {
+    /// Create a new event, with a generated UUID. This function must always
+    /// require and produce the most recent variant of VersionedEvent.
+    pub fn new(
+        operation: Operation,
+        op_extra: Option<EventExtra>,
+        object: String,
+        user: String,
+        occurred_at_unix_epoch_nanos: u64,
+    ) -> Self {
+        Self::new_uuid(
+            Uuid::new_v4(),
+            operation,
+            op_extra,
+            object,
+            user,
+            occurred_at_unix_epoch_nanos,
+        )
+    }
+
+    /// Create a new event. This function must always require and produce the most
+    /// recent variant of VersionedEvent.
+    pub fn new_uuid(
+        uuid: Uuid,
+        operation: Operation,
+        op_extra: Option<EventExtra>,
+        object: String,
+        user: String,
+        occurred_at_unix_epoch_nanos: u64,
+    ) -> Self {
+        let op_extra = op_extra.unwrap_or(EventExtra::None);
+        Self::V1(EventV1::new(
+            uuid,
+            operation,
+            op_extra,
+            object,
+            user,
+            occurred_at_unix_epoch_nanos,
+        ))
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum Operation {
+    CreateCluster,
+    CreateClusterReplica,
+    CreateIndex,
+    CreateSink,
+    CreateSource,
+    CreateView,
+    DropCluster,
+    DropClusterReplica,
+    DropIndex,
+    DropSink,
+    DropSource,
+    DropView,
+    RenameIndex,
+    RenameSink,
+    RenameSource,
+    RenameView,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum EventExtra {
+    None,
+    CreateComputeInstanceReplicaV1(CreateComputeInstanceReplicaV1),
+    DropComputeInstanceReplicaV1(DropComputeInstanceReplicaV1),
+    RenameItemV1(RenameItemV1),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RenameItemV1 {
+    pub previous_name: String,
+    pub new_name: String,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DropComputeInstanceReplicaV1 {
+    pub cluster_name: String,
+    pub replica_name: String,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CreateComputeInstanceReplicaV1 {
+    pub cluster_name: String,
+    pub replica_name: String,
+    pub config: ConcreteComputeInstanceReplicaConfigV1,
```

Comment:
I think we actually just want the logical size in here, since that's the mapping we're going to be putting in customer bills. (They won't necessarily know the mapping between cluster sizes and CPU/memory usage.) Also we should just exempt these remote replicas from audit logging, because they're only available in dev/test environments. You can use `SIZE '1'` in sqllogictest to create a managed replica with a real size.

Reply:
Done.